Add good practices page for secrets (#34291)
kubernetes/website: content/en/docs/concepts/configuration/secret.md

Secrets are similar to {{< glossary_tooltip text="ConfigMaps" term_id="configmap" >}}
but are specifically intended to hold confidential data.

{{< caution >}}
Kubernetes Secrets are, by default, stored unencrypted in the API server's underlying data store (etcd). Anyone with API access can retrieve or modify a Secret, and so can anyone with access to etcd.

> Additionally, anyone who is authorized to create a Pod in a namespace can use that access to read any Secret in that namespace; this includes indirect access such as the ability to create a Deployment.

Reviewer: IMO, this caution is rightly strongly worded and the replacement text is missing many important points that will be lost if they are not covered anywhere else. So we should consider keeping this paragraph as is, on this page because folks would have bookmarked this page and suddenly seeing this "caution" section lost would be a big surprise. 

Please let me if there is a reason I am missing (which is very likely the case) for removing this though.

## Information security for Secrets

Although ConfigMaps and Secrets work similarly, Kubernetes applies some
additional protections to Secret objects.

Secrets often hold values that span a spectrum of importance. Many of these
values can cause escalations within Kubernetes and to external systems. Even if
an individual app can reason about the power of the Secrets it expects to
interact with, other apps within the same namespace can render those assumptions invalid.

A Secret is only sent to a node if a Pod on that node requires it.
For mounting secrets into Pods, the kubelet stores a copy of the data into a

> `tmpfs` so that the confidential data is not written to durable storage.

Reviewer: This can be covered after PR is merged, but it would be good to clarify that "tmpfs" == "in-memory storage" != "stored at rest"

Reviewer 2: Slightly related: if you turn on `NodeSwap` _or_ run the kubelet with the otherwise-unsupported use of a local pagefile, tmpfs data _can_ be stored at rest.

(consider not enabling swap if this behavior isn't what you want; otherwise, encrypt your page files, folks)

When the Pod that depends on the Secret is deleted, the kubelet deletes its
local copy of the confidential data from the Secret.

There may be several containers in a Pod. By default, containers you define
only have access to the default `ServiceAccount` object and its related Secret.
To provide access to other Secret objects, you must explicitly map those objects
to your containers.

There may be Secrets for several Pods on the same node. However, only the
Secrets that a Pod requests are potentially visible within its containers.
One Pod does not have access to the Secrets of another Pod.

{{< warning >}}

> Any privileged containers on a node might have access to all Secrets used

Reviewer: ```suggestion
Any privileged containers on a node can access all Secrets used
```

Author: I'm going to keep as might because we don't know the exact privileges of the containers (the intent wasnt "containers with privileged: true" -- is that what this implies?

Reviewer: Yeah atleast I read it as "privileged: true" :)

Reviewer 3: Might be good to explicitly call this out

```
When a container runs with `privileged: true`, it becomes a privileged container that can access all secrets on the host.
```

## Information security for Secrets

Although ConfigMaps and Secrets work similarly, Kubernetes applies some
additional protections to Secret objects.

Secrets often hold values that span a spectrum of importance. Many of these
values can cause escalations within Kubernetes and to external systems. Even if
an individual app can reason about the power of the Secrets it expects to

> interact with, other apps within the same namespace can render those assumptions invalid.

Reviewer 4: What does this reformatting mean to localization teams? It means they have to compare each and every line of the text because that is how git diff works. As it is shown here, the whole paragraph looks rewritten -- 5 lines removed and 4 lines added. The localization team has to check carefully which word has been changed. The net effect is the   whole paragraph has to be retranslated.

Do we really want to do this, with a compelling reason? Are these changes bring us any value?

Please stop massaging the text in this way. There are tons of topics we have not covered in the docs and there are several hundred open issues to be closed. Could you look into some of those issues instead of reediting the existing content in this way?

Author: You're right, I had this PR out for review before our discussion in Slack and in other PRs about larger changes like this. I'll revert some of the tangential changes that aren't strictly related to this PR and split the commits for the new page to make it easy to see where content was changed vs just moved.

Author: @tengqm reverted most changes to existing Secrets concept (except for one clarity change). Split the creation of the new page into multiple commits so that the diffs are clear about what changed vs what moved. 

@PushkarJ @aramase @ritazh the content changes are identical to the previous state, nothing new to review :)

A Secret is only sent to a node if a Pod on that node requires it.
For mounting secrets into Pods, the kubelet stores a copy of the data into a
`tmpfs` (in-memory storage) so that the confidential data is not written to durable storage. When the Pod that depends on the Secret is deleted, the kubelet deletes its local copy of the confidential data from the Secret.

There may be several containers in a Pod. By default, containers you define
only have access to the default `ServiceAccount` object and its related Secret.
To provide access to other Secret objects, you must explicitly map those objects
to your containers.

There may be Secrets for several Pods on the same node. However, only the
Secrets that a Pod requests are potentially visible within its containers.
One Pod does not have access to the Secrets of another Pod.

{{< warning >}}
Any privileged containers on a node might have access to all Secrets used
on that node.
{{< /warning >}}

For guidelines to manage and improve the security of your Secrets, refer to

> [Good practices for Kubernetes Secrets](/docs/concepts/security/secrets-good-practices).

Reviewer 3: nit: s/Good/Recommended

Reviewer 3: nvm i see the other doc is also "good practices"

Therefore, one Pod does not have access to the Secrets of another Pod.

{{< warning >}}
Any containers that run with `privileged: true` on a node can access all Secrets used
on that node.
{{< /warning >}}

For guidelines to manage and improve the security of your Secrets, refer to

> [Good practices for Kubernetes Secrets](/docs/concepts/security/secrets-good-practices).

Reviewer 2: (nit) @shannonxtreme I'd move this link into the _What's next_ section.

Therefore, one Pod does not have access to the Secrets of another Pod.

{{< warning >}}
Any containers that run with `privileged: true` on a node can access all
Secrets used on that node.
{{< /warning >}}

## {{% heading "whatsnext" %}}

- For guidelines to manage and improve the security of your Secrets, refer to

>   [Good practices for Kubernetes Secrets](/docs/concepts/security/secrets-good-practices).

Reviewer 2: How about putting this link, _and_ a summary of the most salient points, into the caution callout that appears much earlier in the page. See line 32.

(and keep these lines as-is)

Reviewer 2: (@PushkarJ - this is a tweak on your idea, see what you think)

Author: Did so!

Reviewer: That looks good to me @sftim
